File: backend/app/schemas/waitlist.py
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class WaitlistEntry(BaseModel):
    patient_id: str
    visit_type: str
    preferred_locations: list[str] = Field(default_factory=list)
    preferred_time_windows: list[tuple[str, str]] = Field(
        default_factory=list, description="List of preferred HH:MM 24h time windows."
    )
    risk_score: Optional[float] = Field(default=None, ge=0, le=1)
    added_at: datetime


class WaitlistNotification(BaseModel):
    slot_start: datetime
    slot_end: datetime
    location_id: str
    expires_at: datetime
    message: str

File: frontend/src/api/waitlist.ts
import { apiClient } from "./client";

export interface WaitlistEntryPayload {
  patient_id: string;
  visit_type: string;
  preferred_locations?: string[];
  preferred_time_windows?: [string, string][];
  risk_score?: number;
  added_at: string;
}

export interface WaitlistRegisterResponse {
  waitlist_id: string;
}

export interface WaitlistNotification {
  slot_start: string;
  slot_end: string;
  location_id: string;
  expires_at: string;
  message: string;
}

export const registerWaitlist = async (payload: WaitlistEntryPayload) => {
  const { data } = await apiClient.post<WaitlistRegisterResponse>("/waitlist", payload);
  return data;
};

export const requestNotification = async (payload: WaitlistEntryPayload) => {
  const { data } = await apiClient.post<WaitlistNotification>("/waitlist/notify", payload);
  return data;
};
